### ISSUE TITLE
Parameters with bindings are not sorted/ordered correctly.

### ISSUE THREAD

**Issue by mahge** [OpenModelica/OpenModelica#8772]
### Description

Parameters with bindings that depend on other parameters are sometimes not handled correctly. The dependencies of the parameters are evaluated after the them.  

### Steps to Reproduce
test.mo

```mo
package P
  impure function f // function marked impure to prevent the NF from constant evaluating it.
    output Real a[2] = fill(1.0, 2);
    annotation(Inline=false);
  end f;

  function f2
    input Real a[2];
    output Real b[2] = fill(1.0, 2);
    annotation(Inline=false);
  end f2;


  model M
    parameter Real a[2] = f();
    parameter Real d[2] = b;
    parameter Real b[2] = f2(a);
  end M;
end P;
```

test.mos

```mo
loadFile("test.mo");

simulate(P.M);
getErrorString();


val(a[1]);
val(a[2]);

val(b[1]);
val(b[2]);


print("----------------------------\n");

val(d[1]);
val(d[2]);
``` 

output
```sh
1.0
1.0
1.0
1.0
----------------------------

0.0
0.0
```
### Expected Behavior
```sh
1.0
1.0
1.0
1.0
----------------------------

1.0
1.0
```

The values of the array `d` should be set to values of `b` ( = {1.0,1.0}) which is computed by function `f2`. However the assignment of `d := b` is done before the evaluation of `b[2] = f2(a)`

**The input `a[2]` to function `f2` is not used.** However, removing it seems to _hide_ the issue. If that is removed the correct result is obtained. It seems to affect the ordering.

**Reordering the parameter declarations also fixes the problem.** However. it should not be a necessary. This test case is extracted from a bigger model and in the actual case the parameters are ordered properly but the backend resorts them in the wrong order.

### Version and OS

 - OpenModelica Version: v1.19.0-dev-731-g8426dd9ca2
 - OS: Linux

**Comment by mahge** [comment]
@niklwors This test model duplicates your issue. We will continue investigating.

**Comment by lochel** [comment]
Is this a specific issue for array assignments?

**Comment by mahge** [comment]
> Is this a specific issue for array assignments?

I think so but I am not 100% sure. If you simplify this model further the issue will be hidden in most cases. So I am not sure if there is a way to duplicate this using just scalar variables.

I think it might have to do with the fact that we do not expand arrays with function call bindings (`a[:] = f()`) in the NF.

**Comment by perost** [comment]
> > > Is this a specific issue for array assignments?
> > 
> > 
> > I think so but I am not 100% sure. If you simplify this model further the issue will be hidden in most cases. So I am not sure if there is a way to duplicate this using just scalar variables.
> > I think it might have to do with the fact that we do not expand arrays with function call bindings (`a[:] = f()`) in the NF.
> 
> @mahge Should we add Per to this ticket if it has to do with the Frontend?

I see everything anyway. It's not a frontend issue, the only reason the frontend doesn't expand such calls is because we explicitly tell it not to. We added that to avoid the performance issues with your models, since otherwise the function is evaluated once for every array element during simulation. It's not fully supported by the old backend yet though.

**Comment by casella** [comment]
@mahge, I tried to reproduce this with OMEdit, latest nightly, but I get the correct results:
![immagine](https://user-images.githubusercontent.com/4233148/160714415-3fae29f7-89f3-49a0-bc2c-e278b05106cf.png)

Is it perhaps a problem with the scripting environment?

**Comment by mahge** [comment]
@casella I just tried this now and you are right. The NF constant evaluates the function `f`. We had marked the function impure to disable that behavior. I missed that in the test case here. It is now fixed.. 

For OMEdit, you can use this

```mo
package P
  impure function f // function marked impure to prevent NF from evaluating it.
    output Real a[2] = fill(1.0, 2);
    annotation(Inline=false);
  end f;

  function f2
    input Real a[2];
    output Real b[2] = fill(1.0, 2);
    annotation(Inline=false);
  end f2;


  model M
    parameter Real a[2] = f();
    parameter Real d[2] = b;
    parameter Real b[2] = f2(a);
  end M;
end P;
```

Simulated P.M and plot d[1] and d[2].

**Comment by kabdelhak** [comment]
@bernhardbachmann

**Comment by kabdelhak** [comment]
solved with #8933.

The problem was incorrect typing of equations in the backend, now they are correctly parsed as array equations.

I did not test the corresponding real models from BR though. @niklwors @mahge could you do that?

**Comment by niklwors** [comment]
I have tested test models from ticket [#7485](https://github.com/OpenModelica/OpenModelica/issues/7485). There is now a new error for all models. "No system for the symbolic initialization was generated"

**Comment by casella** [comment]
Another likely instance of the same issue is discussed in #8938.

**Comment by casella** [comment]
Yet another probable instance in #8958.

**Comment by casella** [comment]
This should be fixed by #8981, I'll check this with the next nightly build

**Comment by casella** [comment]
I checked with the latest nightly build and all parameters are now correctly assigned the value of 1.

As soon as #8981 is cherry-picked into maintenance/v1.19, as required in #8938, we can also close this one.

**Comment by casella** [comment]
Fixed in #8981 for 1.20.0 and in #9073 for 1.19.2

**Comment by casella** [comment]
Works fine in 1.19.2-dev.beta2
